### PR TITLE
refactor: make http server built flexibly

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

As title, to make other projects able to add extra routes.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/servers/src/grpc/builder.rs [hunk 1]
```diff
@@ -37,12 +37,18 @@ use crate::grpc::prom_query_gateway::PrometheusGatewayService;
 use crate::prometheus_handler::PrometheusHandlerRef;
 use crate::query_handler::OpenTelemetryProtocolHandlerRef;
 
+/// Add a GRPC service (`service`) to a `builder`([RoutesBuilder]).
+/// This macro will automatically add some GRPC properties to the service.
```
Comment (fengjiachun) [inline src/servers/src/grpc/builder.rs:41]:
```suggestion
/// This macro will automatically add some gRPC properties to the service.
```

File: src/servers/src/http.rs [hunk 2]
```diff
@@ -100,25 +100,15 @@ pub static PUBLIC_APIS: [&str; 2] = ["/v1/influxdb/ping", "/v1/influxdb/health"]
 #[derive(Default)]
 pub struct HttpServer {
     // server handlers
```
Comment (fengjiachun) [inline src/servers/src/http.rs:102]: ‌‌Lets's change this comment, there are no handlers now.

File: src/cmd/src/lib.rs [hunk 3]
```diff
@@ -48,6 +48,8 @@ pub trait App {
 pub async fn start_app(mut app: Box<dyn App>) -> error::Result<()> {
     let name = app.name().to_string();
 
+    app.pre_start()?;
```
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/cmd/src/lib.rs:51]: I prefer `pre_start`. `before_start` is longer.

File: src/frontend/src/server.rs [hunk 4]
```diff
@@ -37,16 +37,34 @@ use crate::frontend::{FrontendOptions, TomlSerializable};
...
-    pub fn new(plugins: Plugins) -> Self {
-        Self { plugins }
+impl<T, U> Services<T, U>
+where
+    T: Into<FrontendOptions> + TomlSerializable + Clone,
+    U: FrontendInstance,
+{
+    pub fn new(opts: T, instance: Arc<U>, plugins: Plugins) -> Self {
+        Self {
+            opts,
```
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/frontend/src/server.rs:59]: But the `T` needs to be `TomlSerializable`.

[the remark this answers is not in the document]
Comment (WenyXu) [reply src/frontend/src/server.rs:59]: It seems the opts always be convert into `FrontendOptions` and use the `TomlSerializable` implement of it? There is anything I missed?

File: src/cmd/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/cmd/src/lib.rs:51]: Or `prestart`? `pre_start` looks a little oddly

File: src/frontend/src/server.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/frontend/src/server.rs:59]: The `T` can be `MixOptions`, too.

File: src/cmd/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/cmd/src/lib.rs:51]: Still prefer `pre_start`. "pre" "start" is more clear than "prestart", the latter one can be easily seem as "p restart" or "prest art" or else.

**Review by fengys1996 — APPROVED** [review]
LGTM

**Review by WenyXu — APPROVED** [review]

**Pull request merged**